bdk_electrum_example/src/main.rs

```diff
@@ -0,0 +1,277 @@
+/// The [`Update`] returned contains invalidation if any.
+/// The function also updates the [`TxGraph`] internally as [`Tx`] data is not returned
+pub fn fetch_update_from_client(
+    tx_graph: &mut TxGraph,
+    sparse_chain: &SparseChain,
+    client: &Client,
+    scripts: &mut impl Iterator<Item = (u32, Script)>,
+    stop_gap: u32,
+) -> Result<(Option<u32>, Update)> {
+    let (last_valid, invalidate) = detect_reorg(sparse_chain, client)?;
+
+    let existing_txid = sparse_chain
+        .iter_txids()
+        .map(|(conf_ht, txid)| {
+            (
+                txid,
+                conf_ht.map_or(TxHeight::Unconfirmed, |ht| TxHeight::Confirmed(ht)),
+            )
+        })
+        .collect::<HashSet<_>>();
+
+    let mut last_active_index = None;
+    let mut unused_script_count = 0u32;
+    let mut script_history_txid = HashSet::new();
+
+    loop {
+        let (index, script) = scripts.next().expect("its an infinite iterator");
+
+        let history = client
+            .script_get_history(&script)?
+            .iter()
+            .map(|history_result| {
+                if history_result.height > 0 {
+                    return (
+                        history_result.tx_hash,
+                        TxHeight::Confirmed(history_result.height as u32),
+                    );
+                } else {
+                    return (history_result.tx_hash, TxHeight::Unconfirmed);
+                };
+            })
+            .collect::<HashSet<_>>();
+
+        if history.is_empty() {
+            unused_script_count += 1;
+        } else {
+            last_active_index = Some(index);
+            script_history_txid.extend(history.iter());
+        }
+
+        if unused_script_count >= stop_gap {
+            break;
+        }
+    }
+
+    let txids_to_update: HashSet<_> = script_history_txid
+        .difference(&existing_txid)
+        .cloned()
+        .collect();
+
+    for tx in client.batch_transaction_get(
+        txids_to_update
+            .iter()
+            .map(|(txid, _)| txid)
+            .collect::<Vec<_>>(),
+    )? {
+        if tx_graph.insert_tx(&tx) {
+            debug!("added new wallet tx {} in the TxGraph", tx.txid());
+        }
+    }
+
+    // TODO: Is there a better way to get latest tip without subscribing?
+    let new_tip = client
+        .block_headers_subscribe()
+        .map(|header_resp| BlockId {
+            height: header_resp.height as u32,
+            hash: header_resp.header.block_hash(),
+        })?;
+
+    Ok((
+        last_active_index,
+        Update {
+            txids: txids_to_update
+                .into_iter()
+                .collect::<HashMap<Txid, TxHeight>>(),
+            last_valid,
+            invalidate,
+            new_tip,
+        },
+    ))
+}
+
+pub fn perform_electrum_sync<K: Debug + Clone + Ord>(
+    client: &Client,
+    tracker: &mut KeychainTracker<K>,
+    chain: &mut SparseChain,
+    graph: &mut TxGraph,
```

Comment:
In the examples we don't want to have these things be mutable while doing IO. The point we want to make is that you can get stuff from the wallet and then drop mutable references, fetch data without any mutable references open and then later get a mutable reference to apply the result. I know for CLI we don't even need to do this because there's only one thread but let's see if we can structure stuff so it's like this. I know the example I wrote originally doesn't do this but see if you can get this message across when designing it.

Reply:
Okay.. Ya I think I see what you are saying.. So we will have some "applier" to insert data inside this structs?